File: agent/runner_factory.py
from __future__ import annotations

from typing import Any

from .model_router import ModelRoute
from .runner import AgentRunner
from .telemetry import TokenTracker
from .tools import ToolRegistry


def build_routed_runner(
    *,
    route: ModelRoute,
    registry: ToolRegistry,
    system_prompt: str,
    token_tracker: TokenTracker | None,
    usage_type: str,
    max_tokens_cap: int | None = None,
    memory_store=None,
    compactor=None,
    todo_store=None,
    control_manager=None,
    max_context: int | None = None,
    max_turns: int = 12,
) -> AgentRunner:
    snapshot = route.snapshot
    fallback = route.fallback
    max_tokens = snapshot.generation.max_tokens
    if max_tokens_cap is not None:
        max_tokens = min(max_tokens_cap, max_tokens)
    kwargs: dict[str, Any] = {}
    if max_context is not None:
        kwargs["max_context"] = max_context
    return AgentRunner(
        provider=snapshot.provider,
        model=snapshot.model,
        registry=registry,
        system_prompt=system_prompt,
        max_tokens=max_tokens,
        temperature=snapshot.generation.temperature,
        reasoning_effort=snapshot.generation.reasoning_effort,
        provider_name=snapshot.provider_name,
        model_role=snapshot.model_role,
        route_reason=route.reason,
        route_estimated_tokens=route.estimated_tokens,
        fallback_provider=fallback.provider if fallback else None,
        fallback_model=fallback.model if fallback else None,
        fallback_provider_name=fallback.provider_name if fallback else None,
        fallback_generation=fallback.generation if fallback else None,
        fallback_model_role=fallback.model_role if fallback else "main",
        usage_type=usage_type,
        memory_store=memory_store,
        token_tracker=token_tracker,
        compactor=compactor,
        todo_store=todo_store,
        control_manager=control_manager,
        max_turns=max_turns,
        **kwargs,
    )

File: packages/core/src/agent/runner-factory.ts
/**
 * runner_factory (MIG-CORE-010)。对齐 Python `agent/runner_factory.py:build_routed_runner`。
 * 按单一激活模型构造 AgentRunner，供主 Agent、子代理和 Team 复用。
 */
import type { LLMProvider } from '../providers/base'
import type { ModelRoute } from '../model/router'
import type { PromptContextPlan, PromptSectionInput } from '../prompts/manifest'
import type { ToolRegistry } from '../tools/registry'
import {
  AgentRunner,
  type CompactorLike,
  type ControlManagerRunnerHost,
  type AgentRunnerHookHost,
  type MemoryStoreLike,
  type TodoStoreLike,
  type TokenTrackerLike,
} from './runner'
import type { RunnerGoalRecordingHost } from './runner-goal-recording'
import type { GoalContextProvider } from '../context/pipeline'
import type { GoalToolHost } from '../goals/tools'

export function buildRoutedRunner(opts: {
  route: ModelRoute
  registry: ToolRegistry
  systemPrompt: string
  tokenTracker: TokenTrackerLike | null
  usageType: string
  maxTokensCap?: number | null
  memoryStore?: MemoryStoreLike | null
  compactor?: CompactorLike | null
  todoStore?: TodoStoreLike | null
  controlManager?: ControlManagerRunnerHost | null
  maxContext?: number | null
  maxTurns?: number
  workspaceRoot?: string | null
  promptSections?: PromptSectionInput[] | null
  promptContextPlan?: PromptContextPlan | null
  promptSnapshotDir?: string | null
  sessionId?: string | null
  streamingToolExecution?: boolean
  hooks?: AgentRunnerHookHost | null
  goalObservationRecorder?: RunnerGoalRecordingHost | null
  goalToolHost?: Pick<GoalToolHost, 'visibleToolNames'> | null
  goalContextProvider?: GoalContextProvider | null
  goalContextHint?:
    | (() => Promise<{
        readonly goalId: string
        readonly lastEventSeq: number
      } | null>)
    | null
  onGoalCompacted?: (() => void) | null
}): AgentRunner {
  const snapshot = opts.route.snapshot
  let maxTokens = snapshot.generation.maxTokens
  if (opts.maxTokensCap !== null && opts.maxTokensCap !== undefined) {
    maxTokens = Math.min(opts.maxTokensCap, maxTokens)
  }
  return new AgentRunner({
    provider: snapshot.provider as unknown as LLMProvider,
    model: snapshot.model,
    registry: opts.registry,
    systemPrompt: opts.systemPrompt,
    maxTokens,
    temperature: snapshot.generation.temperature,
    reasoningEffort: snapshot.generation.reasoningEffort,
    providerName: snapshot.providerName,
    modelEntryId: snapshot.modelEntryId ?? snapshot.entryName,
    supportsToolCall: snapshot.profile?.toolCall ?? true,
    routeReason: opts.route.reason,
    routeEstimatedTokens: opts.route.estimatedTokens,
    usageType: opts.usageType,
    memoryStore: opts.memoryStore ?? null,
    tokenTracker: opts.tokenTracker,
    compactor: opts.compactor ?? null,
    todoStore: opts.todoStore ?? null,
    controlManager: opts.controlManager ?? null,
    maxContext:
      opts.maxContext ??
      snapshot.profile?.contextWindowTokens ??
      snapshot.contextWindowTokens,
    maxTurns: opts.maxTurns ?? 12,
    workspaceRoot: opts.workspaceRoot ?? null,
    promptSections: opts.promptSections ?? null,
    promptContextPlan: opts.promptContextPlan ?? null,
    promptSnapshotDir: opts.promptSnapshotDir ?? null,
    sessionId: opts.sessionId ?? null,
    streamingToolExecution: opts.streamingToolExecution ?? false,
    hooks: opts.hooks ?? null,
    goalObservationRecorder: opts.goalObservationRecorder ?? null,
    goalToolHost: opts.goalToolHost ?? null,
    goalContextProvider: opts.goalContextProvider ?? null,
    goalContextHint: opts.goalContextHint ?? null,
    onGoalCompacted: opts.onGoalCompacted ?? null,
  })
}
